### PR TITLE
DEEP-10726: Support for deep license scanning

### PR DESCRIPTION
**Pull request by hrithik-777**
## Support for deep License Scanning 
- Added generic Recursive walker for deep license scanning.
- Modified the filtering of File system for PostAnalyze API.
- Added LicenseV2 field in Package struct which is used for storing concluded licenses
- Note: Licenses array is used for declared licenses, licensesV2 is for concluded licenses.
- Added license scan related fields and configs in trivy options.
- Extracting license text from the file content using license classifier

### CODE REVIEW

File: pkg/fanal/analyzer/analyzer.go [hunk 1]
```diff
@@ -56,7 +56,14 @@ type SecretScannerOption struct {
 
 type LicenseScannerOption struct {
 	// Use license classifier to get better results though the classification is expensive.
-	Full                      bool
+	// Flag indicates whether license scanner is enabled or not
+	Enabled bool
```
Comment (theakshaymenon) [inline pkg/fanal/analyzer/analyzer.go:61]: Why cant we use the `Full` scan bool here?

File: pkg/scanner/local/scan.go [hunk 2]
```diff
@@ -389,14 +389,16 @@ func (s Scanner) scanLicenses(target types.ScanTarget, options types.ScanOptions
...
-				category, severity := scanner.Scan(license)
+			for _, license := range lib.LicensesV2 {
+				category, severity := scanner.Scan(license.Name)
 				langLicenses = append(langLicenses, types.DetectedLicense{
-					Severity:   severity,
-					Category:   category,
-					PkgName:    lib.Name,
-					Name:       license,
-					Confidence: 1.0,
+					Severity:    severity,
```
Comment (theakshaymenon) [inline pkg/scanner/local/scan.go]: We need the pkgVersion to link the license to the pkg

File: pkg/fanal/analyzer/analyzer.go (hunk 1 above)
Comment (hrithik-777) [reply pkg/fanal/analyzer/analyzer.go:61]: Full corresponds to --license-full flag
Enabled -> --scanners license

So just to validate properly, only when both the flags are passed, deep license scan is enabled for us.

File: pkg/scanner/local/scan.go (hunk 2 above)
Comment (hrithik-777) [reply pkg/scanner/local/scan.go]: sure

File: go.mod [hunk 3]
```diff
@@ -2,6 +2,10 @@ module github.com/deepfactor-io/trivy
 
 go 1.21
```
Comment (theakshaymenon) [inline go.mod]: Use go 1.21 here

File: pkg/fanal/analyzer/analyzer.go (hunk 1 above)
Comment (namandf) [reply pkg/fanal/analyzer/analyzer.go:61]: But how are both different? --license-full vs --scanners

File: pkg/utils/split.go [hunk 4]
```diff
@@ -11,3 +14,92 @@ var typesForPkgSplit = []ftypes.TargetType{ftypes.NodePkg, ftypes.Yarn, ftypes.N
...
+		}
+	}
+
+	return lo.Uniq(cleanedParts)
+}
+
+// validate whether given license is SPDX defined or not
+func ValidateLicense(license string) bool {
+	ok, _ := spdxexp.ValidateLicenses([]string{license})
+	return license != "" && ok
```
Comment (john-d8r) [inline pkg/utils/split.go]: Early return if license == ""

File: pkg/utils/split.go [hunk 5]
```diff
@@ -11,3 +14,92 @@ var typesForPkgSplit = []ftypes.TargetType{ftypes.NodePkg, ftypes.Yarn, ftypes.N
...
+	for _, license := range licenses {
+		for _, license := range splitLicense(license) {
+			ok, _ := spdxexp.ValidateLicenses([]string{license})
+			foundSPDXLicense = foundSPDXLicense || ok
+
+			if ok {
+				result = append(result, license)
+			}
+		}
+	}
```
Comment (john-d8r) [inline pkg/utils/split.go]: We could combine these loops

File: pkg/types/report.go [hunk 6]
```diff
@@ -78,6 +78,9 @@ const (
 	FormatSPDXJSON   Format = "spdx-json"
 	FormatGitHub     Format = "github"
 	FormatCosignVuln Format = "cosign-vuln"
+
+	LicenseTargetOSPkg       string = "OS Packages"
```
Comment (john-d8r) [inline pkg/types/report.go:82]: How about `os-pkg-license` and `loose-file-license`

Comment (john-d8r) [reply pkg/types/report.go:82]: Never mind

File: pkg/sbom/core/bom.go [hunk 7]
```diff
@@ -126,6 +127,10 @@ type Component struct {
 	// SPDX: package.licenseConcluded, package.licenseDeclared
 	Licenses []string
 
+	// LicensesV2 is a list of concluded licenses that apply to the component
+	// SPDX: package.licenseConcluded
+	LicensesV2 []types.License
```
Comment (john-d8r) [inline pkg/sbom/core/bom.go]: Any reason for naming it LIcensesV2 and not ConcludedLicenses ?

Comment (hrithik-777) [reply pkg/sbom/core/bom.go]: earlier i used it store declared + concluded licenses as part of licesnses v2, but now we only store concluded licenses in it. we can change if it's needed, what do you say? @john-d8r @theakshaymenon

File: pkg/licensing/classifier.go [hunk 8]
```diff
@@ -84,3 +103,226 @@ func Classify(filePath string, r io.Reader, confidenceLevel float64) (*types.Lic
...
+		}
+
+		licenseLink := fmt.Sprintf("https://spdx.org/licenses/%s.html", match.Name)
+
+		// extract & get the checksum and license text for the current match
+		textChecksum, licenseText := extractLicenseText(input.Content, match.StartLine, match.EndLine)
+
+		// persist the license text in given folder
+		if input.LicenseTextCacheDir != "" && len(licenseText) > 0 {
+			go input.persistLicenseText(textChecksum, licenseText)
```
Comment (namandf) [inline pkg/licensing/classifier.go:199]: won't this fail silently in case of error and impact scan result?

File: pkg/licensing/classifier.go [hunk 9]
```diff
@@ -84,3 +103,226 @@ func Classify(filePath string, r io.Reader, confidenceLevel float64) (*types.Lic
...
+			return fmt.Errorf("failed to persist license text. (checksum: %s, err: %s)", checksum, err.Error())
+		}
+	}
+
+	return nil
+}
+
+// extracts the copyright text from given license text
+func extractCopyrightFromLicenseText(text string) string {
+	re := regexp.MustCompile(`(?i)Copyright\s+((\(c\)|©)?\s*\d{4}(?:-\d{4})?\s+.*?)(?:\\n|$)`)
```
Comment (namandf) [inline pkg/licensing/classifier.go]: Would it make sense to do this once?

File: pkg/licensing/classifier.go [hunk 10]
```diff
@@ -84,3 +103,226 @@ func Classify(filePath string, r io.Reader, confidenceLevel float64) (*types.Lic
...
+		errClassifierPoolInit = fmt.Errorf("failed to initialize license classifier pool, err: %s", err.Error())
+		return errClassifierPoolInit
+	}
+
+	return nil
+}
+
+// gets a classifier from the pool of classifiers
+func (p *licenseClassifierPool) Get() *classifier.Classifier {
+	p.mutex.Lock()
```
Comment (namandf) [inline pkg/licensing/classifier.go]: What's the use of lock here? Isn't the channel communication already blocking? i.e if somebody triggers Get, they'll get a classifier only if available in a channel , else it will wait for it

File: pkg/licensing/classifier.go [hunk 11]
```diff
@@ -84,3 +103,226 @@ func Classify(filePath string, r io.Reader, confidenceLevel float64) (*types.Lic
...
+
+func (input *ClassifierInput) persistLicenseText(
+	checksum string,
+	licenseText string,
+) error {
+	// <checksum>.txt would be the name of the file where license text is persisted
+	licenseTextFilepath := filepath.Join(input.LicenseTextCacheDir, fmt.Sprintf("%s.txt", checksum))
+
+	// persist only when file does not exist
+	if _, err := os.Stat(licenseTextFilepath); errors.Is(err, fs.ErrNotExist) {
```
Comment (namandf) [inline pkg/licensing/classifier.go]: you can instead check if the file exists and return. Rest of the code can then be outside the if block

File: go.mod [hunk 12]
```diff
@@ -4,6 +4,8 @@ go 1.22.0
 
 toolchain go1.22.4
 
+replace github.com/deepfactor-io/go-dep-parser => github.com/deepfactor-io/go-dep-parser v0.0.0-20240522114955-c3b72fb971a8
```
Comment (theakshaymenon) [inline go.mod]: go-dep-parser isn't  required anymore right?

File: go.mod [hunk 13]
```diff
@@ -130,7 +132,11 @@ require (
 	sigs.k8s.io/yaml v1.4.0
 )
 
-require github.com/alicebob/miniredis/v2 v2.33.0
+require (
+	github.com/alicebob/miniredis/v2 v2.33.0
+	github.com/deepfactor-io/go-dep-parser v0.0.0-00010101000000-000000000000
```
Comment (theakshaymenon) [inline go.mod]: remove go-dep-parser

File: pkg/dependency/parser/utils/utils.go [hunk 14]
```diff
@@ -81,3 +81,7 @@ func MergeMaps(parent, child map[string]string) map[string]string {
 	}
 	return newParent
 }
+
+func PackageID(name, version string) string {
```
Comment (theakshaymenon) [inline pkg/dependency/parser/utils/utils.go]: Can we use this function instead? dependency.ID 
https://github.com/deepfactor-io/trivy/blob/232bb0974e0db5e770ba843392440a838c43dca5/pkg/dependency/id.go#L16

File: pkg/utils/fsutils/license_scanning_fs.go [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+
+	foundPackageDependencyDir := checkPackageDependencyDir(fsys, root, w.PackageDependencyDir)
+	if foundPackageDependencyDir {
+		return w.handlePackageDependencies(fsys, root, pkgID)
+	}
+
+	return true, nil
+}
+
+// For special packages (like scoped packages in npm), we need to recurse further for scanning
```
Comment (namandf) [inline pkg/utils/fsutils/license_scanning_fs.go:179]: please elaborate and may be add some example. Its difficult to comprehend

File: pkg/utils/fsutils/license_scanning_fs.go [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+			w.mutex.Lock()
+			w.licenses[pkg.PackageID()] = append(w.licenses[pkg.PackageID()], pkg.DeclaredLicenses()...)
+			w.mutex.Unlock()
+
+			return pkg.PackageID(), nil
+		}
+	}
+
+	var pkgID string
+	if parentPkgID == "" {
```
Comment (namandf) [inline pkg/utils/fsutils/license_scanning_fs.go]: this is confusing and doesn't seem right. Same variable returns a string constant and a package id (name/version), which have a totally different meaning.

File: pkg/utils/fsutils/license_scanning_fs.go [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+			w.Logger.Debug("pkgID is already present, skipping recursive walk",
+				log.String("pkgID", pkgID), log.String("path", root))
+			return true, nil
+		} else {
+			w.pkgIDMap.Store(pkgID, struct{}{})
+		}
+	}
+
+	required := func(filePath string, d fs.DirEntry) bool {
+		// Skip PkgDependency Dir (Ex: node_modules) directory and for Package manifest file
```
Comment (namandf) [inline pkg/utils/fsutils/license_scanning_fs.go]: please elaborate. Its difficult to comprehend

File: pkg/fanal/analyzer/language/dotnet/nuget/nuget.go [hunk 18]
```diff
@@ -41,22 +40,41 @@ type nugetLibraryAnalyzer struct {
...
-	foundLicenses := make(map[string][]string)
+	foundLicenses := make(map[string][]types.License)
+
+	var looseLicenses []types.LicenseFile
 
-	// We saved only config and lock files in the FS,
+	// We save only config and lock files in the filtered FS,
 	// so we need to parse all saved files
+	// Note: If deep license scan is enabled, we'll save every file in the FS,
+	// so exquiste check is required here
```
Comment (namandf) [inline pkg/fanal/analyzer/language/dotnet/nuget/nuget.go]: nit pick. Do you mean `explicit`?

File: pkg/licensing/classifier.go (hunk 8 above)
Comment (hrithik-777) [reply pkg/licensing/classifier.go:199]: how will it impact scan result?

File: pkg/licensing/classifier.go (hunk 10 above)
Comment (hrithik-777) [reply pkg/licensing/classifier.go]: yeah, i'm removing it. It's not needed.

File: pkg/licensing/classifier.go (hunk 9 above)
Comment (hrithik-777) [reply pkg/licensing/classifier.go]: right, i'm changing it.

File: pkg/utils/fsutils/license_scanning_fs.go (hunk 16 above)
Comment (hrithik-777) [reply pkg/utils/fsutils/license_scanning_fs.go]: I agree. changing it, please check

File: go.mod (hunk 3 above)
Comment (hrithik-777) [reply go.mod]: after latest trivy sync, it's 1.22

File: go.mod (hunk 12 above)
Comment (hrithik-777) [reply go.mod]: yes, removing it.

File: go.mod (hunk 13 above)
Comment (hrithik-777) [reply go.mod]: yes, removing it.

File: pkg/dependency/parser/utils/utils.go (hunk 14 above)
Comment (hrithik-777) [reply pkg/dependency/parser/utils/utils.go]: yes using it, please check

File: pkg/utils/split.go [hunk 19]
```diff
@@ -11,3 +14,95 @@ var typesForPkgSplit = []ftypes.TargetType{ftypes.NodePkg, ftypes.Yarn, ftypes.N
...
+	var result []string
+	var tempLicenses []string
+
+	// Split and validate licenses
+	for _, license := range licenses {
+		// skip empty licenses
+		if len(license) == 0 {
+			continue
+		}
+		tempLicenses = append(tempLicenses, license)
```
Comment (namandf) [inline pkg/utils/split.go:33]: why is this required?

File: pkg/utils/split.go [hunk 20]
```diff
@@ -11,3 +14,95 @@ var typesForPkgSplit = []ftypes.TargetType{ftypes.NodePkg, ftypes.Yarn, ftypes.N
...
+
+	// distinct license names will be returned
+	return lo.Uniq(result)
+}
+
+// Func parses the given license name.
+// If it's a single license, return it. If it's combination of licenses, split it based on the seperators and return it
+func splitLicense(license string) []string {
+	// common separators present in license name
+	separators := []string{" OR ", " AND "}
```
Comment (namandf) [inline pkg/utils/split.go:58]: Is there a specific standard followed? If not, it might be better to use regexes to extract substrings/parts

File: pkg/utils/fsutils/license_scanning_fs.go [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+	pkgID, foundPkgManifest, err := w.processPackageManifest(fsys, root)
+	if err != nil {
+		w.Logger.Error("Failed to process package manifest", log.String("Err", err.Error()))
+		return false, err
+	}
+
+	if !foundPkgManifest {
+		if parentPkgID == "" {
+			w.Logger.Debug("Parent PkgID is empty. Adding to loose licenses", log.String("path", root))
+			pkgID = types.LOOSE_LICENSES
```
Comment (namandf) [inline pkg/utils/fsutils/license_scanning_fs.go:119]: i was under the impression that this was fixed. We discussed that using the same variable for 2 different types of values is confusing

File: pkg/utils/split.go (hunk 20 above)
Comment (hrithik-777) [reply pkg/utils/split.go:58]: I see that this OR and AND seperators are most commonly used when declaring licenses. Is regex processing required?

File: pkg/utils/split.go (hunk 19 above)
Comment (hrithik-777) [reply pkg/utils/split.go:33]: i'm updating the license to this temp licenses array. This is to filter out any empty license strings present 
 
`licenses = tempLicenses
	if !foundSPDXLicense {
		return lo.Uniq(licenses)
	}`

File: pkg/fanal/analyzer/language/dotnet/nuget/nuget.go (hunk 18 above)
Comment (hrithik-777) [reply pkg/fanal/analyzer/language/dotnet/nuget/nuget.go]: I''ll update

File: pkg/utils/split.go (hunk 19 above)
Comment (namandf) [reply pkg/utils/split.go:33]: Hmm. Can't this be handle when we create the licenses array?

File: pkg/utils/split.go (hunk 20 above)
Comment (namandf) [reply pkg/utils/split.go:58]: There might be cases where there is an extra space etc. We'll miss those. If we have a regex then those cases can also be handled right?

File: pkg/utils/split.go (hunk 19 above)
Comment (hrithik-777) [reply pkg/utils/split.go:33]: Can be but it's more effort. Since licenses are populated for different languages in different packages, i need to modify those as well. (i.e not just node, dotnet related things). for now, i'm handling them in post processing. @namandf
